[ADD] get pins script

gfcapalbo opened the pull request.

gfcapalbo commented:
@thomaspaulb i put this in WIP again.
I need to support when "target:"  is used.
we usually don;'t use it , because jan made a MR where target is not mandatory anymore (the first merge is the base branch) / but i saw a couple of waft instances where it is used so i want to add support/

gfcapalbo commented:
@thomaspaulb Refactor done and added support for pinning target:

example: 
      `. get_pins --only-base=True`

gives:

`OCA/bank-payment:
  defaults:
    depth: 2000
  remotes:
    origin: https://github.com/OCA/bank-payment.git
    wpichler: https://github.com/wpichler/bank-payment.git
    acsone: https://github.com/acsone/bank-payment.git
  target: origin ${ODOO_VERSION} 69083802dfcfa20ea8ed1a6f94b6abe868c32c5c
  merges:
  - origin ${ODOO_VERSION}
  - wpichler 16.0-mig-account_payment_order
  - acsone 16.0-mig-account_payment_mode
`

While `. get_pins`  gives:

`OCA/bank-payment:
  defaults:
    depth: 2000
  remotes:
    origin: https://github.com/OCA/bank-payment.git
    wpichler: https://github.com/wpichler/bank-payment.git
    acsone: https://github.com/acsone/bank-payment.git
  target: origin ${ODOO_VERSION} 69083802dfcfa20ea8ed1a6f94b6abe868c32c5c
  merges:
  - origin ${ODOO_VERSION} 69083802dfcfa20ea8ed1a6f94b6abe868c32c5c
  - wpichler 16.0-mig-account_payment_order 6e7bfc43ac767c4bfef4ce4bd5bb5e614fec3050
  - acsone 16.0-mig-account_payment_mode de70eee76f47c18e72585520ad43d91f0ba2ee13
`

NL66278 commented on bin/get_pins:
```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python
+# Version: v.22.05.30
+# -*- coding: utf-9 -*-
```
utf-9 ??? Anyway not needed, supposing we use python > 3.0

gfcapalbo replied:
vim accident.

NL66278 commented:
@gfcapalbo Please use:
```
if __name__ == "__main__":
    sys.exit(main())
```
And a Usage function as described here: https://www.artima.com/weblogs/viewpost.jsp?thread=4829

gfcapalbo commented:
@NL66278  added support for imposing the correct merge, and followed through your comments/


Diff, you can see that where the necessary depth merge exceeds the waft_depth_merge variable it is replaced by the correct value + 25. ( I added 25 because if we used the precise value, by the time we deploy the new yaml , new commits may have already been added, making that value insufficient)



![image](https://user-images.githubusercontent.com/8994879/236280792-38266274-1794-4089-b5c1-124640dc03a7.png)

thomaspaulb commented on bin/get_pins:
```diff
@@ -0,0 +1,268 @@
+#!/usr/bin/env python
```
In order to make this slightly easier to call, without having to activate the venv yourself, I propose to:

- Rename this script to `get_pins.py` and remove executable permissions
- Add an executable bash script `get_pins`, which just acts as an alias and basically contains the same boilerplate as all the other scripts do:

```
SCRIPT_PATH="$(cd "$(/usr/bin/dirname "${0}")" && /bin/pwd)"
ODOO_WORK_DIR="${SCRIPT_PATH}"
. "${ODOO_WORK_DIR}/.env-default" && \
. "${ODOO_WORK_DIR}/.env-shared" && \
. "${ODOO_WORK_DIR}/.env-secret"
cd "${ODOO_WORK_DIR}"

if [ "$(id -u)" -eq 0 ]; then
  /bin/echo "Please dont use sudo command with script"
  exit 1
fi

. .venv/bin/activate && python waftlib/bin/get_pins.py $@
```

It's not that elegant but it's handy to use, and hopefully in the future we'll refactor this to something better

thomaspaulb replied:
(Then while you're at it, could you rebase on latest master and do the same for Danny's `translate-module` script?)

thomaspaulb commented on bin/get_pins:
```diff
@@ -0,0 +1,268 @@
+        .stdout.decode("utf-8")
+        .replace("\n", "")
+    )
+    return int(mindepth)
```
When I run this for Freshfilter as a test, on my local, I get this error:

```
(.venv) antiflu@qqq:/data/buildout/fresh14$ waftlib/bin/get_pins 
Traceback (most recent call last):
  File "waftlib/bin/get_pins", line 265, in <module>
    main()
  File "waftlib/bin/get_pins", line 240, in main
    repo_path,
  File "waftlib/bin/get_pins", line 146, in process_depth
    return int(mindepth)
ValueError: invalid literal for int() with base 10: ''
```

Could you check out what's going on there? You can just clone the Freshfilter test build on your local, then checkout this branch in waftlib

gfcapalbo replied:
This happens when a merge of "type 2 " exists:

 <remote> <PIN>        merge_type=2       


No branchname, just the commit. I have seen this syntax used. 
Need to manage the calculation of depth for these cases.

gfcapalbo replied:
Solved, verified with FF

thomaspaulb commented on bin/get_pins:
```diff
@@ -0,0 +1,268 @@
+
+
+def get_branchname(splitted_merge):
+    return splitted_merge[1].replace("${ODOO_VERSION}", ODOO_VERSION)
```
This seems a bit crude since we allow any kind of variable in repos.yaml, not only `ODOO_VERSION`. Since we're already loading all the `.env-*`  files anyway, can we just support any variable by leveraging [Template.substitute](https://docs.python.org/3/library/string.html#template-strings) in the way that [gitaggregator does in these 4 lines](https://github.com/acsone/git-aggregator/blob/master/git_aggregator/config.py#L164-L167)?

gfcapalbo replied:
Can't do for all variables. WAFT_DEPTH_MERGE will often not be in the env at all, when it is not in the env WAFT defaults it to 100. so i have to support the many cases where it is not present in env.

thomaspaulb replied:
Hey but that's not true, maybe it's because the branch is not rebased on latest master, but [.env-default](https://github.com/sunflowerit/waftlib/blob/master/templates/14.0/.env-default#L45) is guaranteed to be there and always contains a default value for `WAFT_DEPTH_MERGE`. So if you load `.env-default`, which you do, you can depend on it containing that variable.

thomaspaulb commented on bin/get_pins:
```diff
@@ -0,0 +1,268 @@
+    """
+    with open(REPOS_YAML) as yaml_file:
+        for doc in yaml.safe_load_all(yaml_file):
+            for repo in doc:
```
I'd like to add some verbosity here, to avoid the user thinking that nothing's happening or the program is waiting for keyboard input, or something. Maybe just "processing OCA/web.. found pin 123456... changing to abcdef..."

thomaspaulb reviewed (changes requested):
Review

gfcapalbo commented:
@thomaspaulb 

 could not reply directly in threads.... some bug.
Added verbosity:

......now it notifies: 
==> processing repo R
     ===>  processing merge X
                   Updating pin,  inserting new pin,  increasing depth merge from xx to xy


Also replacing all ENV variables as suggested, that means all WAFT_DEPTH_MERGE and all WAFT_DEPTH_DEFAULT are replaced too, is that okay?

![Screenshot from 2023-05-09 13-47-30](https://github.com/sunflowerit/waftlib/assets/8994879/b626f143-1f33-4ac6-a12f-60530c249d01)

gfcapalbo commented:
@thomaspaulb ALl requests done. I will make Danny's translation script execution  in a separate branch.

thomaspaulb commented on get_pins:
```diff
@@ -0,0 +1,15 @@
+ODOO_WORK_DIR="${SCRIPT_PATH}"
+. "${ODOO_WORK_DIR}/.env-default" ; \
+. "${ODOO_WORK_DIR}/.env-shared" ; \
+. "${ODOO_WORK_DIR}/.env-secret"; cd "${ODOO_WORK_DIR}"
```
If I now run the script I get:

```
antiflu@qqq:/data/buildout/fresh14$ waftlib/get_pins 
waftlib/get_pins: 5: .: Can't open /mnt/data/buildout/fresh14/waftlib/.env-default
```

`ODOO_WORK_DIR="${SCRIPT_PATH}"` works in the case of `translate-module` because we're expecting `./bootstrap` to [symlink](https://github.com/sunflowerit/waftlib/blob/master/bootstrap#L285) the script into the root folder, and it's then run from there.

However, in the case of [reset-password](https://github.com/sunflowerit/waftlib/blob/master/bin/reset-password#L14), it's expected to be run as `waftlib/bin/reset-password` so there we add `../..` to `ODOO_WORK_DIR` and then it works like that.

I prefer the latter so that we don't overcrowd the root folder, so:

1. Can you move `get_pins` to `bin` folder
2. Can you add `../..` to the end of line 4 here
3. Maybe add a README line explaining that you can update the pins by running `waftlib/bin/get_pins`. If there's currently no line about `reset-password` feel free to add that as well

And then it should work properly.

thomaspaulb commented on bin/get_pins.py:
```diff
@@ -0,0 +1,292 @@
+    """
+    with open(REPOS_YAML) as yaml_file:
+        yaml_substituted = Template(yaml_file.read())
+        yaml_substituted = yaml_substituted.substitute(os.environ)
```
Okay wait... now it substitutes all the environment variables, then loads the YAML, and then saves, with all the variables substituted, which is why you'll lose `WAFT_DEPTH_DEFAULT` which will be replaced by `1`, and so on.

That's not what we want, I think.

What we want is the YAML to be loaded and saved with the variable names intact; just that whenever we do something with for example a branch name or a depth value, and we need the actual value, that's when we should do the variable subsitution thing in order to extract that value (eg in `get_branchname` function) but also when doing `waft_depth = doc[repo]["defaults"].get("depth")`.

I guess what you can do is to create a helper function `decode_variables` which then does this `Template().substitue` thing on the string that you pass it. Then, you can call it inside of the loop below, for branch name and depth where needed, but you don't have to do it on the whole YAML which would cause loss of information.

thomaspaulb reviewed (changes requested):
Some comments still

gfcapalbo commented:
@thomaspaulb done all suggestions:

     - Made executable and updated Readme
     - Made variable substitution only per-line when needed , no information lost when no subst needed 
     - rebased on more current master.
     - tested in 2 different wafts.
     - ran black.

ddejong-therp commented:
Hey giovanni, maybe it would be nice if the script has the execution permission.
